### ISSUE TITLE
Deep MNIST for Experts

### ISSUE THREAD

**Issue by killelder** [killelder/tensorflownote#4]
https://www.tensorflow.org/get_started/mnist/pros

--------------------------------------------------------------
https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/examples/tutorials/mnist/mnist_softmax.py
一樣參考上面這一段code
這一part是想辦法提高accuracy

tensorflow 依賴C++高效率的運算能力, session就是連結圖和計算
InteractiveSession 這段看不懂, 上網google以後說interactiveSession主要是可以少key一點code
python code的腳色是建構計算圖, 然後等計算的結果

-------------------------------------------------------
Build a Multilayer Convolutional Network
在這之前都跟前一part一樣
想看的人可以直接參考
https://github.com/killelder/tensorflownote/issues/3

這裡會使用convolutional neural network (CNN, 卷積神經網路)
可以大幅提高準確率
因為打算使用ReLu 當activation, 所以必須要讓weight不是0

ex. 1-1
def weight_variable(shape):
  initial = tf.truncated_normal(shape, stddev=0.1)
  return tf.Variable(initial)

def bias_variable(shape):
  initial = tf.constant(0.1, shape=shape)
  return tf.Variable(initial)

tensorflow也給予很多彈性在convolution和pooling,
如何處理邊界和stride size(一次卷積的size)怎麼定義
下面使用1當作stride size, 而且使用"same" padding, 會在外圍的部分補0, 以至於output跟input size一樣
pooling使用 max pooling over 2x2 blocks
ex. 1-2
def conv2d(x, W):
  return tf.nn.conv2d(x, W, strides=[1, 1, 1, 1], padding='SAME')

def max_pool_2x2(x):
  return tf.nn.max_pool(x, ksize=[1, 2, 2, 1],
                        strides=[1, 2, 2, 1], padding='SAME')

First Convolutional Layer(第一層convolution layer)
這一層希望對5x5的patch分出32種features
weight tensor必須要shape of [5,5,1,32], 前面兩個是patch size, 第三個是input channels, 最後一個是output channels
bias tensor就直接是output channels
把原本x reshape to a 4d tensor [-1, 28, 28, 1]) [幾個sample, width, height, color channel]
ex. 1-3
W_conv1 = weight_variable([5, 5, 1, 32])
b_conv1 = bias_variable([32])
x_image = tf.reshape(x, [-1, 28, 28, 1])
h_conv1 = tf.nn.relu(conv2d(x_image, W_conv1) + b_conv1)
h_pool1 = max_pool_2x2(h_conv1)

Second Convolutional Layer(第二層convolution layer)
第二層64 features of 5x5 patch
ex. 1-4
W_conv2 = weight_variable([5, 5, 32, 64])
b_conv2 = bias_variable([64])

h_conv2 = tf.nn.relu(conv2d(h_pool1, W_conv2) + b_conv2)
h_pool2 = max_pool_2x2(h_conv2)

Densely Connected Layer
這時image size被reduced to 7x7 
28x28 經過maxpooling ==> 14x14 再一次maxpooling ==> 7x7
convolution 不會減少size, 但是會把每一個位置都做相同的是給出features
我們採用1024 neurons 去展開整個[7*7*64]的image

ex. 1-5
h_pool2_flat = tf.reshape(h_pool2, [-1, 7*7*64])
W_fc1 = weight_variable([7 * 7 * 64, 1024])
b_fc1 = bias_variable([1024])
h_fc1 = tf.nn.relu(tf.matmul(h_pool2_flat, W_fc1) + b_fc1)

Dopout
為了避免overfitting, 我們加入dropout, before the readout layer,
我們建立一個keep_prob是為了, 當訓練的時候可以加dropout, 當test的時候可以不要dropout

ex. 1-6
keep_prob = tf.placeholder(tf.float32)
h_fc1_drop = tf.nn.dropout(h_fc1, keep_prob)

Readout Layer

ex. 1-7
W_fc2 = weight_variable([1024, 10])
b_fc2 = bias_variable([10])
y_conv = tf.matmul(h_fc1_drop, W_fc2) + b_fc2

跟之前簡單的模型不同的點有3個
1. 利用ADAM optimizer
2. keep_prob 來控制dropout rate
3. add log every 100th iteration in the training process
選擇使用tf.Session, 因為更好的分割創造圖的階段與執行圖的階段
且當離開with的時候Session就會關閉

ex. 1-8
cross_entropy = tf.reduce_mean(
    tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y_conv))
train_step = tf.train.AdamOptimizer(1e-4).minimize(cross_entropy)
correct_prediction = tf.equal(tf.argmax(y_conv, 1), tf.argmax(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())
  for i in range(20000):
    batch = mnist.train.next_batch(50)
    if i % 100 == 0:
      train_accuracy = accuracy.eval(feed_dict={
          x: batch[0], y_: batch[1], keep_prob: 1.0})
      print('step %d, training accuracy %g' % (i, train_accuracy))
    train_step.run(feed_dict={x: batch[0], y_: batch[1], keep_prob: 0.5})

  print('test accuracy %g' % accuracy.eval(feed_dict={
      x: mnist.test.images, y_: mnist.test.labels, keep_prob: 1.0}))

!!dropout雖然對抗overfitting很有用, 但她通常在訓練大型neural network才有用
